# Exploring eBay Car Sales Data
For this project, we're using a modified version of a German eBay website dataset (found [here](https://data.world/data-society/used-cars-data)), revolving around used cars.

DataQuest modified the dataset in two ways:
- They sampled 50.000 data points from the full set to ensure the code runs smoothly in our hosted environment
- They dirtied the dataset a little to encourage practice with data cleaning

The dataset includes many columns, so for posteriosity's sake, a data dictionary:

| Column Name          | Description |
|----------------------|-------------|
| dateCrawled          | When this ad was first crawled. All field-values are taken from this date. |
| name                 | Name of the car. |
| seller               | Whether the seller is private or a dealer. |
| offerType            | The type of listing. |
| price                | The price on the ad to sell the car. |
| abtest               | Whether the listing is included in an A/B test. |
| vehicleType          | The type of vehicle. |
| yearOfRegistration   | The year in which the car was first registered. |
| gearbox              | The type of transmission. |
| powerPS              | The power of the car in PS. |
| model                | The car model name. |
| odometer             | How many kilometers the car has driven. |
| monthOfRegistration  | The month in which the car was first registered. |
| fuelType             | What type of fuel the car uses. |
| brand                | The brand of the car. |
| notRepairedDamage    | If the car has a damage which is not yet repaired. |
| dateCreated          | The date the eBay listing was created. |
| nrOfPictures         | The number of pictures in the ad. |
| postalCode           | The postal code for the location of the vehicle. |
| lastSeenOnline       | When the crawler saw this ad last online. |

## Importing libraries and the dataset

Now, first we'll import the necessary libraries, read the CSV and convert it into a DataFrame.

In [1]:
import pandas as pd
import numpy as np
autos = pd.read_csv("eBay car sales.csv", encoding="latin1")

Reading the CSV without specifying an encoding method (and thus relying on the standard `UTF-8`), will yield a `UnicodeDecodeError`. Attempting using one of the next most popular encodings, `Latin-1`, solved the error. 

## Exploring the data
Let's also take a look at one of Jupyter Notebook's nifty features: a quick render of the dataset's first and last few values of a pandas object. We'll do this by inserting the object (`autos` in this case) into a separate cell, and then running said cell.

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

To get a better idea of the data's properties (i.e. shape, object types), we'll also subject the pandas object `autos` to the `DataFrame.info()` method, and `DataFrame.describe()`.

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [4]:
autos.describe()

yearOfRegistration       powerPS  monthOfRegistration  nrOfPictures  \
count        50000.000000  50000.000000         50000.000000       50000.0   
mean          2005.073280    116.355920             5.723360           0.0   
std            105.712813    209.216627             3.711984           0.0   
min           1000.000000      0.000000             0.000000           0.0   
25%           1999.000000     70.000000             3.000000           0.0   
50%           2003.000000    105.000000             6.000000           0.0   
75%           2008.000000    150.000000             9.000000           0.0   
max           9999.000000  17700.000000            12.000000           0.0   

         postalCode  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

Based on the above, we can come to the following conclusions:
- The DataFrame has **20 columns**.
- The column `name` uses **snake_case** for its values, while the column labels themselves use **CamelCase** (which is harder to read). Taken together, this reads as inconsistent.
- The following columns are likely **missing entries**: `vehicleType`, `gearbox`, `model`, `fuelType`, `notRepairedDamage`. None have more than 20% (20% of 5000 equals 5000*0.2 = 1000) `Null` values, though.
- The `dateCrawled`, `dateCreated` and `lastSeen` columns contain strings; we may want to convert these to `DateTime` or `pandas Timestamp` objects, so we can use pandas' built-in tools to work with times and dates.
- Meanwhile, the `yearOfRegistration` and `monthOfRegistration` contain `integers`. Though at first instinct we might want to try converting these to `DateTime` or `pandas TimeStamp` objects, both require full dates (day, month, year) to work properly. As such, these are best left as integers.
- The `price` column contains **string values**, instead of floats; the `odometer`column likewise contains string values, instead of integers.
- The currency used is also **dollars**, instead of euros (which is one would expect from products sold in Germany). However, considering that we don't know whether this is an already-converted value (from euros to dollars), or if the dollar sign is an automatic addition to all values in that column without further altering it, we won't convert the values.
- Many of the string values are in **German** – which might not be terribly useful to those who don't speak a word of German, necessitating translation into English.

## Data cleaning
### Converting camelCase to snake_case in column labels
Unfortunately, it's not a simple matter of replacing white spaces with underscores. Let's first see what columns exactly require new labels:

In [5]:
# Viewing the names of the columns
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

Though it's not a major issue, for readability's and brevity's sake, we'll want to change the following column names to snake_case:
- `dateCrawled`to `date_crawled`
- `offerType`to `offer_type`
- `abtest`to `ab_test`
- `vehicleType` to `vehicle_type`
- `yearOfRegistration`to `registration_year`
- `powerPS` to `horsepower`
- `monthOfRegistration`to `registration_month`
- `fuelType` to `fuel_type`
- `notRepairedDamage` to `unrepaired_damage`
- `dateCreated` to `ad_created`
- `nrOfPictures`to `number_of_pictures`
- `postalCode` to `postal_code`
- `lastSeen` to `ad_last_seen`

To do this, we'll simply paste the list of existing column names and modify the names as stated, assigning that list to the variable `new_autos_columns`. Then we'll assign the modified list of column names back to the `DataFrame.columns` attribute.

In [6]:
# Renaming the columns to make them more descriptive
new_autos_columns = ["date_crawled", "name", "seller", "offer_type", "price", "ab_test",
       "vehicle_type", "registration_year", "gearbox", "horsepower", "model",
       "odometer", "registration_month", "fuel_type", "brand",
       "unrepaired_damage", "ad_created", "number_of_pictures", "postal_code",
       "ad_last_seen"]

# Assigning the altered column names back to the columns attribute
autos.columns = new_autos_columns

# Verifying the current state of the autos DataFrame
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  horsepower   model   odometer  registration_month fuel_type  \
0    manuell         158  andere  150,000km                   3       lpg   
1  automatik         286     7er  150,000km                   6    benzin   
2    manuell         102    golf   70,000km                   7    benzin   
3  automatik          71  fortwo   70,000km                   6    benzin   
4    manuell           0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  number_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00                   0   
1         bmw              nein  2016-04-04 00:00:00                   0   
2  volkswagen              nein  2016-03-26 00:00:00                   0   
3       smart              nein  2016-03-12 00:00:00                   0   
4        ford              nein  2016-04-01 00:00:00                   0   

   postal_code         ad_last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Now all the column names are similarly formatted, while clearly identifying what data they contain and following the conventional approach while working with Python. This will make them easy to work with during the upcoming analyses.

### Identifying + dropping columns with largely identical values
To identify whether there are any columns that feature largely of exclusively identical values, we'll primarily use two methods: `DataFrame.describe()` and `Series.value_counts()`.

In [7]:
# Getting descriptive statistics – and including non-numeric columns in our output
autos.describe(include="all")

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox    horsepower  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  number_of_pictures   postal_code  \
count                 50000             50000.0  50000.000000   
unique                   76                 NaN           NaN   
top     2016-04-03 00:00:00                 NaN           NaN   
freq                   1946                 NaN           NaN   
mean                    NaN                 0.0  50813.627300   
std                     NaN                 0.0  25779.747957   
min                     NaN                 0.0   1067.000000   
25%                     NaN                 0.0  30451.000000   
50%                     NaN                 0.0  49577.000000   
75%                     NaN                 0.0  71540.000000   
max                     NaN                 0.0  99998.000000   

               ad_last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

Looking at the above descriptives, we can conclude that the following columns are eligible to be dropped due to containing only one or two types of variables, rendering their contributions to future analyses minimal: `seller`, `offer_type`, `ab_test`, `gearbox`, and `unrepaired_damage`.

Furthermore, the minimum and maximum values for the column `number_of_pictures` are both zero, implying that this column only contains zeros. Let's verify that the majority of these columns' values are, indeed, of one specific type, rather than equally divided:

In [8]:
# Defining what columns we want to parse in the loop below
columns = ["seller", "offer_type", "ab_test", "gearbox", "unrepaired_damage", "number_of_pictures"]

# Looping through the columns and printing the value counts for each
for column in columns:
    print(autos[column].value_counts())

seller
privat        49999
gewerblich        1
Name: count, dtype: int64
offer_type
Angebot    49999
Gesuch         1
Name: count, dtype: int64
ab_test
test       25756
control    24244
Name: count, dtype: int64
gearbox
manuell      36993
automatik    10327
Name: count, dtype: int64
unrepaired_damage
nein    35232
ja       4939
Name: count, dtype: int64
number_of_pictures
0    50000
Name: count, dtype: int64


This confirms that the data in the columns `seller`, `offer_type` and `number_of_pictures` contain largely, if not exclusively, a single value. Given that these values are unlikely to provide us with helpful data during future analyses, we'll drop them from the DataFrame. We'll do this using the `DataFrame.drop()` method.

In [9]:
autos.drop(columns=["seller", "offer_type", "number_of_pictures"], inplace=True)

# Resetting the index to prevent row counting issues in the future
autos.reset_index(drop=True, inplace=True)

# Confirming that the columns were dropped
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   price               50000 non-null  object
 3   ab_test             50000 non-null  object
 4   vehicle_type        44905 non-null  object
 5   registration_year   50000 non-null  int64 
 6   gearbox             47320 non-null  object
 7   horsepower          50000 non-null  int64 
 8   model               47242 non-null  object
 9   odometer            50000 non-null  object
 10  registration_month  50000 non-null  int64 
 11  fuel_type           45518 non-null  object
 12  brand               50000 non-null  object
 13  unrepaired_damage   40171 non-null  object
 14  ad_created          50000 non-null  object
 15  postal_code         50000 non-null  int64 
 16  ad_last_seen        50

### Converting numeric data from strings to floats
Based on the above table and our initial data exploration, we'll also want to clean up the `price` and `odometer` columns by:

- Deleting non-numeric characters
- Converting the columns' contents to numeric dtypes
- Renaming the `odometer` label in specific to `odometer_km`.

We'll remove the '$' characters and remove the commas before converting the values of the `price` column to a `float64` dtype. For the `odometer` column, we'll remove 'km' and any commas before likewise converting the values to floats.

In [10]:
# Removing dollar signs and commas from the price column
clean_price = autos["price"].str.strip("$").str.replace(",", "")
# Converting the price column"s values to float64
clean_price = clean_price.astype(float)
# Assigning clean_price back to autos["price"]
autos["price"] = clean_price

In [11]:
# Removing "km" and commas from the odometer column
clean_odometer = autos["odometer"].str.replace("km", "").str.replace(",", "")
# Converting the odometer column"s values to floats
clean_odometer = clean_odometer.astype(float)
# Assigning clean_odometer back to autos["odometer"]
autos["odometer"] = clean_odometer
# Renaming the odometer column
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)

Let's take a look at the data in its most recent form to determine what steps we must take next:

In [12]:
autos.describe(include="all")

date_crawled         name         price ab_test vehicle_type  \
count                 50000        50000  5.000000e+04   50000        44905   
unique                48213        38754           NaN       2            8   
top     2016-04-02 11:37:04  Ford_Fiesta           NaN    test    limousine   
freq                      3           78           NaN   25756        12859   
mean                    NaN          NaN  9.840044e+03     NaN          NaN   
std                     NaN          NaN  4.811044e+05     NaN          NaN   
min                     NaN          NaN  0.000000e+00     NaN          NaN   
25%                     NaN          NaN  1.100000e+03     NaN          NaN   
50%                     NaN          NaN  2.950000e+03     NaN          NaN   
75%                     NaN          NaN  7.200000e+03     NaN          NaN   
max                     NaN          NaN  1.000000e+08     NaN          NaN   

        registration_year  gearbox    horsepower  model    odometer_km  \
count        50000.000000    47320  50000.000000  47242   50000.000000   
unique                NaN        2           NaN    245            NaN   
top                   NaN  manuell           NaN   golf            NaN   
freq                  NaN    36993           NaN   4024            NaN   
mean          2005.073280      NaN    116.355920    NaN  125732.700000   
std            105.712813      NaN    209.216627    NaN   40042.211706   
min           1000.000000      NaN      0.000000    NaN    5000.000000   
25%           1999.000000      NaN     70.000000    NaN  125000.000000   
50%           2003.000000      NaN    105.000000    NaN  150000.000000   
75%           2008.000000      NaN    150.000000    NaN  150000.000000   
max           9999.000000      NaN  17700.000000    NaN  150000.000000   

        registration_month fuel_type       brand unrepaired_damage  \
count         50000.000000     45518       50000             40171   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     30107       10687             35232   
mean              5.723360       NaN         NaN               NaN   
std               3.711984       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created   postal_code         ad_last_seen  
count                 50000  50000.000000                50000  
unique                   76           NaN                39481  
top     2016-04-03 00:00:00           NaN  2016-04-07 06:17:27  
freq                   1946           NaN                    8  
mean                    NaN  50813.627300                  NaN  
std                     NaN  25779.747957                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  30451.000000                  NaN  
50%                     NaN  49577.000000                  NaN  
75%                     NaN  71540.000000                  NaN  
max                     NaN  99998.000000                  NaN

### Other columns that require cleaning


This leaves us with the several columns that still require attention.

The following columns likely contain incorrect **outliers** that had best be filtered out:
- The `price` column might contain inaccurate values, with an unlike minimum price of $0, and a maximum price of $99999999.
- The `registration_year` column likewise contains several unlikely values, like the year 1000 and 9999.
- The `horsepower` column has a questionable minimum value of 0, and a maximum value of 17700.
- The `odometer_km` column shows the same value (150000) on the 50th and 75th percentiles and max values.
- The `registration_month` column contains a minimum value of 0, which does not correspond to a month at all.

We still have to tackle the conversion from strings to proper `DateTime` objects:
- `date_crawled`
- `ad_created`
- `ad_last_seen`

The following columns contain **missing values** that should be addressed:
- `vehicle_type`
- `gearbox`
- `model`
- `fuel_type`

On top of that, we still need to **translate** the German string values into English.

### Removing outliers
In the columns `price`, `registration_year`, `horsepower`, `odometer_km` and `registration_month`, we seem to have some unlikely values. Let's explore these columns further and decide whether to remove the rows with outliers or not.

To remove outliers in which quantiles have any meaning (numerical, not categorical), we'll first create a function that'll efficiently allow us to remove rows with outliers from specified columns, and will return a DataFrame with outliers removed from the specified column.

*Edit: turns out we only needed this exactly _once_, but for future reference, we'll leave it here and use it as needed!

In [13]:
def remove_outliers(df, column, lower_percentile=0.01, upper_percentile=0.99):
    """
    Removes outliers from a DataFrame column, based on the specified percentiles.

    Parameters:
    df: pandas DataFrame
    column: name of the column from which we want to remove outliers, as a string
    lower_percentile: lower percentile threshold, as a float
    upper_percentile: upper percentile threshold, as a float

    Returns:
    pandas DataFrame with outliers removed from the specified column
    """
    # Setting the lower and upper bounds for the specified column
    lower_bound, upper_bound = df[column].quantile([lower_percentile, upper_percentile])
    # Removing outliers from the specified column
    cleaned_df = df[df[column].between(lower_bound, upper_bound)]
    # Assigning the cleaned DataFrame back to the original DataFrame
    return cleaned_df

Let's also create a function that will seamlessly allow us to explore the data of the columns we want to explore.

In [14]:
def explore_column(df, column_name):
    """
    Explores a specified column in a pandas DataFrame.

    Parameters:
    - df: pandas DataFrame.
    - column_name: String name of the column to explore.

    Prints the shape of unique values, description, and value counts sorted in ascending order for the specified column.
    """
    print(f"Exploring the {column_name} column – assessing the numbers of unique values, description, and value counts")
    print("Unique shape: \n", df[column_name].unique().shape)
    print("Description: \n", df[column_name].describe())
    print("Value counts, sorted ascending: \n", df[column_name].value_counts().sort_index(ascending=True))
    print("-----------------------------------------------------------------------------------")

# Example usage:
# explore_column(autos, "odometer_km")

#### Assessing and removing `price` outliers

To prvent scientific notation, we'll first format all floats to display four decimals. We'll go with four instead of two, because later analyses will require that measure of precision to let us come to precise conclusions about the data's distribution.

In [15]:
# Setting the display option to show four decimal places instead of scientific notation
pd.set_option("display.float_format", "{:.4f}".format)
# Exploring the price column – assessing the numbers of unique values, description and value counts
explore_column(autos, "price")

Exploring the price column – assessing the numbers of unique values, description, and value counts
Unique shape: 
 (2357,)
Description: 
 count      50000.0000
mean        9840.0438
std       481104.3805
min            0.0000
25%         1100.0000
50%         2950.0000
75%         7200.0000
max     99999999.0000
Name: price, dtype: float64
Value counts, sorted ascending: 
 price
0.0000           1421
1.0000            156
2.0000              3
3.0000              1
5.0000              2
                 ... 
10000000.0000       1
11111111.0000       2
12345678.0000       3
27322222.0000       1
99999999.0000       1
Name: count, Length: 2357, dtype: int64
-----------------------------------------------------------------------------------


Based on the above, it seems that 1421 cars were given away for free – but you can't list items for free on eBay. Furthermore, the maximum price (99999999.00) is far above the 75% percentile (7200.00).

Removing all the zero values will remove approximately a third of the values in this column. However, given that meaningless values like these in such large numbers warp the accuracy of the data immensely, we'll remove them (alongside the high-price outliers).

To do this as efficiently as possible, we'll delete any values below the 1st percentile (lower bound) and above the 99th percentile (upper bound), using our earlier defined function: `remove_outliers`.

In [16]:
autos = remove_outliers(autos, "price")

# Checking the new value counts of the price column
print("After removing price outliers: \n", autos["price"].value_counts().sort_index(ascending=True))

After removing price outliers: 
 price
0.0000        1421
1.0000         156
2.0000           3
3.0000           1
5.0000           2
              ... 
35700.0000       1
35790.0000       2
35800.0000       1
35890.0000       2
35900.0000       9
Name: count, Length: 2081, dtype: int64


Now the maximum price is $35900, a much more reasonable maximum value for a (secondhand) car.

#### Assessing and removing `registration_year` outliers

In [17]:
# Exploring the registration_year column – assessing the numbers of unique values, description and value counts
explore_column(autos, "registration_year")

Exploring the registration_year column – assessing the numbers of unique values, description, and value counts
Unique shape: 
 (94,)
Description: 
 count   49502.0000
mean     2004.9652
std       104.5489
min      1000.0000
25%      1999.0000
50%      2003.0000
75%      2008.0000
max      9999.0000
Name: registration_year, dtype: float64
Value counts, sorted ascending: 
 registration_year
1000    1
1001    1
1111    1
1500    1
1800    2
       ..
5911    1
8888    1
9000    2
9996    1
9999    4
Name: count, Length: 94, dtype: int64
-----------------------------------------------------------------------------------


Cars did not exist in the year 1000 (it seems the first modern car as we know it was [registered around 1886](https://arc.net/l/quote/uoskjnvb) by what we know as Mercedes-Benz).

Let's see what is the most recent possible `registration_year` by seeing when the data was crawled. We'll do this, because the original dataset has since been removed, so we can't glean any information from the corresponding Kaggle page.

In [18]:
# Using the max method to find the most recent crawl date
autos["date_crawled"].max()

'2016-04-07 14:36:56'

Since the most recent crawl date was in 2016, no values above 2016 should remain in the `registration_year` column. Let's fix that, using the `between()` method once more.

In [19]:
# Using the between method to see how many values fall outside the 1886-2020 range
autos["registration_year"].between(1886, 2016).value_counts()

registration_year
True     47539
False     1963
Name: count, dtype: int64

1443 values are not between our specified years, which makes up about 1443/49502 ≈ 3% . We could remove the entire rows in which these values are found.

However, for the sake of preserving the other information in those rows, we'll just remove any values from the `registration_year` column from before 1885, and after 2020 (since at the time of writing, it is 2024). We'll use the `between()` method to accomplish this.

In [20]:
# Using the between method to remove outliers from the registration_year column
autos = autos[autos["registration_year"].between(1886, 2016)]

# Verifying if all outliers were removed
print("Value count and description after outlier removal: \n", autos["registration_year"].between(1886, 2016).value_counts())
autos["registration_year"].describe()

Value count and description after outlier removal: 
 registration_year
True    47539
Name: count, dtype: int64


count   47539.0000
mean     2002.7579
std         7.2221
min      1910.0000
25%      1999.0000
50%      2003.0000
75%      2007.0000
max      2016.0000
Name: registration_year, dtype: float64

Now it's clear that among the listed cars, the earliest year of registration is 1927, and the latest is 2016. Further more, at least 75% of the cars (≈ 31919, rounded) were first registrered in or after 1999, making making 75% of the cars listed on eBay Germany during the crawling period 20 - 21 years old at most. 

Meanwhile, only 25% (≈ 10639) of the cars is no older than 11 - 12 years. The newer the car, the harder it'll be to find!

#### Assessing and removing `horsepower` outliers

In [21]:
# Exploring the horsepower column – assessing the numbers of unique values, description and value counts
explore_column(autos, "horsepower")

Exploring the horsepower column – assessing the numbers of unique values, description, and value counts
Unique shape: 
 (414,)
Description: 
 count   47539.0000
mean      115.2780
std       194.8751
min         0.0000
25%        71.0000
50%       105.0000
75%       150.0000
max     17700.0000
Name: horsepower, dtype: float64
Value counts, sorted ascending: 
 horsepower
0        4966
1           4
2           2
3           2
4           4
         ... 
14009       1
15001       1
15016       1
16312       1
17700       1
Name: count, Length: 414, dtype: int64
-----------------------------------------------------------------------------------


Though no working car has 0 horsepower, it could be a placeholder for people who simply didn't know and/or for people who sold cars that were out of working order (and therefore, technically, had 0 horsepower). Nevertheless, we don't particularly want 5460 likely placeholder values dirtying our DataFrame. 

Since there are some extremely old cars that might have only 1 horsepower, we'll go the cautious route and accept that as our lower boundary.

Likewise, the highest horsepower values are unlikely to hold true: per 2020, the most powerful car apparently had 'only' [around 2000 horsepower](https://www.topgear.com/car-news/dubai-motor-show/2012bhp-aspark-owl-most-powerful-production-hypercar). As such, we'll take 2000 as our `upper_boundary` in our boolean mask.

*Such as the first Mercedes-Benz, see earlier link!

In [22]:
# Using the between method to remove outliers from the horsepower column
autos = autos[autos["horsepower"].between(1, 2000)]

# Verifying if all outliers were removed
print("Value count and description after outlier removal: \n", autos["horsepower"].between(1, 2000).value_counts())
autos["horsepower"].describe()

Value count and description after outlier removal: 
 horsepower
True    42558
Name: count, dtype: int64


count   42558.0000
mean      125.6454
std        69.7062
min         1.0000
25%        80.0000
50%       116.0000
75%       150.0000
max      1998.0000
Name: horsepower, dtype: float64

Now the lowest amount of horsepower is 1, and the highest is 1998. That is in line with what we might expect from cars first registered between 1927 and 2020.

#### Assessing and removing `odometer_km` outliers

In [23]:
# Exploring the odometer_km column – assessing the numbers of unique values, description and value counts
explore_column(autos, "odometer_km")

Exploring the odometer_km column – assessing the numbers of unique values, description, and value counts
Unique shape: 
 (13,)
Description: 
 count    42558.0000
mean    126242.0696
std      38799.6318
min       5000.0000
25%     125000.0000
50%     150000.0000
75%     150000.0000
max     150000.0000
Name: odometer_km, dtype: float64
Value counts, sorted ascending: 
 odometer_km
5000.0000        525
10000.0000       187
20000.0000       633
30000.0000       677
40000.0000       717
50000.0000       927
60000.0000      1044
70000.0000      1114
80000.0000      1285
90000.0000      1543
100000.0000     1882
125000.0000     4559
150000.0000    27465
Name: count, dtype: int64
-----------------------------------------------------------------------------------


The minimum value of 5000 km seem reasonable, as does the maximum value of 150000 km. The number of cars with 50000 km on their odometer (approximately 57%), however, does seem a little excessive. However, perhaps eBay doesn't allow values over 150000 to be entered, or, upon downloading the CSV, values of '150000+' were merged with those of '150000'.

In a similar vein: we might have expected more than merely 13 unique values for this column, but a reasonable reason for this is that this is the result of rounding options on eBay's listing page.

As there don't appear to be any particular outliers in this column, we'll leave it as is and move on to the next.

#### Assessing and removing `registration_month` outliers

In [24]:
# Exploring the registration_month column – assessing the numbers of unique values, description and value counts
explore_column(autos, "registration_month")

Exploring the registration_month column – assessing the numbers of unique values, description, and value counts
Unique shape: 
 (13,)
Description: 
 count   42558.0000
mean        5.9622
std         3.5955
min         0.0000
25%         3.0000
50%         6.0000
75%         9.0000
max        12.0000
Name: registration_month, dtype: float64
Value counts, sorted ascending: 
 registration_month
0     2732
1     2881
2     2696
3     4496
4     3647
5     3663
6     3831
7     3458
8     2807
9     3049
10    3300
11    2988
12    3010
Name: count, dtype: int64
-----------------------------------------------------------------------------------


The first interesting finding, is that there are 13 unique values, while a year only has 12 months. It seems the suspicious value here is '0'. This could be a typo (people wanting to type '10' for October, or maybe '01' for January'), or errant programmers using zero-indexing while counting the months.

It could also be a case of 0 being used as a placeholder for cars with an unknown registration month.

Now, we _could_ remove these rows to be rid of any placeholder data that might skew our analysis. Given that about 8% of thecurrent number of rows (already over 10% less than when we started) has this value, however, we'll not remove these rows, instead bearing in mind the presence of placeholder values should the need arise.

## Translating German values into English
To increase the legibility of the dataset's columns with string values for those who don't speak German, it's a good idea to translate them. Let's take a look at all the string columns first.

In [25]:
# Viewing all columns with string ("object") dtypes
print(autos.select_dtypes(include="object"))

              date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       ab_test vehicle_type

There's no need for us to translate the values in the `name` column, as this is column is of no consequence to any likely-to-be-performed analyses. Meanwhile, the `ab_test` column contains English terms. Here are the columns we need to further investigate and translate:

- `vehicle_type`
- `gearbox`
- `model`
- `fuel_type`
- `brand` (just in case there is a word meaning 'other' or 'unidentified' in there)
- `unrepaired_damage`

Translations will happen based on my own knowledge of German (as someone who grew up around the language), verified by means of Wikipedia articles.

Let's check out the unique values of each column, and translate what needs to be translated. We'll use `values_count()`, because I personally find the outcome easier to read than that of `unique()`.

### Translating the `vehicle_type` column

In [26]:
# Exploring unique values in the vehicle_type column
autos["vehicle_type"].value_counts()

vehicle_type
limousine     11902
kleinwagen     9703
kombi          8445
bus            3738
cabrio         2766
coupe          2265
suv            1776
andere          328
Name: count, dtype: int64

Words that need translating:
- _limousine_ ⇔ sedan
- _kleinwagen_ ⇔ supermini, or subcompact
- _kombi_ ⇔ station wagon
- _bus_ ⇔ van
- _cabrio_ ⇔ convertible, or cabriolet
- _andere_ ⇔ other

Let's go with the first option there where we have two, and get to replacing the values in this column!

In [27]:
# Creating a mapping dictionary to convert German vehicle types to English
vehicle_type_mapping = {
    "limousine": "sedan",
    "kleinwagen": "supermini",
    "kombi": "station wagon",
    "bus": "van",
    "cabrio": "convertible",
    "coupe": "coupe",
    "suv": "suv",
    "andere": "other"
}

# Mapping the German vehicle types to English
autos["vehicle_type"] = autos["vehicle_type"].map(vehicle_type_mapping)

# Verifying the changes
autos["vehicle_type"].value_counts()

vehicle_type
sedan            11902
supermini         9703
station wagon     8445
van               3738
convertible       2766
coupe             2265
suv               1776
other              328
Name: count, dtype: int64

### Translating the `gearbox` column

In [28]:
# Exploring unique values in the gearbox column
autos["gearbox"].value_counts()

gearbox
manuell      32726
automatik     9069
Name: count, dtype: int64

Let's translate both values here:
- _manuell_ ⇔ manual
- _automatik_ ⇔ automatic

In [29]:
# Creating a mapping dictionary to convert German gearbox types to English
gearbox_mapping = {
    "manuell": "manual",
    "automatik": "automatic"
}

# Mapping the German gearbox types to English
autos["gearbox"] = autos["gearbox"].map(gearbox_mapping)

# Verifying the changes
autos["gearbox"].value_counts()

gearbox
manual       32726
automatic     9069
Name: count, dtype: int64

### Translating the `fuel_type` column

In [30]:
# Exploring unique values in the fuel_type column
autos["fuel_type"].value_counts()

fuel_type
benzin     26464
diesel     13047
lpg          624
cng           65
hybrid        35
elektro       14
andere         9
Name: count, dtype: int64

We'll translate the following terms:
- _benzin_ ⇔ gasoline, or petrol (we'll go with _gasoline_)
- _elektro_ ⇔ electric
- _andere_ ⇔ other

In [31]:
# Creating a mapping dictionary to convert German fuel types to English
fuel_type_mapping = {
    "benzin": "gasoline",
    "diesel": "diesel",
    "lpg": "lpg",
    "cng": "cng",
    "hybrid": "hybrid",
    "elektro": "electric",
    "andere": "other"
}

# Mapping the German fuel types to English
autos["fuel_type"] = autos["fuel_type"].map(fuel_type_mapping)

# Verifying the changes
autos["fuel_type"].value_counts()

fuel_type
gasoline    26464
diesel      13047
lpg           624
cng            65
hybrid         35
electric       14
other           9
Name: count, dtype: int64

### Translating the `brand` column

In [32]:
# Exploring unique values in the brand column
autos["brand"].value_counts()

brand
volkswagen        9136
bmw               4846
opel              4576
mercedes_benz     4046
audi              3755
ford              2923
renault           1927
peugeot           1272
fiat              1072
seat               796
skoda              724
mazda              660
nissan             631
citroen            590
smart              590
toyota             547
hyundai            434
volvo              406
mini               399
mitsubishi         349
honda              346
sonstige_autos     316
kia                310
alfa_romeo         288
suzuki             254
chevrolet          235
chrysler           157
porsche            141
dacia              114
daihatsu            99
subaru              98
jeep                93
land_rover          74
saab                72
jaguar              64
daewoo              55
rover               51
trabant             46
lancia              45
lada                21
Name: count, dtype: int64

From the above, we only need to translate _sonstige autos_ into _other cars_. Because the list of available brands is as long as it is, we'll use the `replace()` method to save time and space.

In [33]:
# Using the replace method to turn "sonstige_autos" into "other cars"
autos["brand"] = autos["brand"].replace("sonstige_autos", "other cars")

# Verifying the changes
autos["brand"].value_counts()

brand
volkswagen       9136
bmw              4846
opel             4576
mercedes_benz    4046
audi             3755
ford             2923
renault          1927
peugeot          1272
fiat             1072
seat              796
skoda             724
mazda             660
nissan            631
citroen           590
smart             590
toyota            547
hyundai           434
volvo             406
mini              399
mitsubishi        349
honda             346
other cars        316
kia               310
alfa_romeo        288
suzuki            254
chevrolet         235
chrysler          157
porsche           141
dacia             114
daihatsu           99
subaru             98
jeep               93
land_rover         74
saab               72
jaguar             64
daewoo             55
rover              51
trabant            46
lancia             45
lada               21
Name: count, dtype: int64

### Translating the `unrepaired_damage` column

In [34]:
# Exploring unique values in the unrepaired_damage column
autos["unrepaired_damage"].value_counts()

unrepaired_damage
nein    32071
ja       4197
Name: count, dtype: int64

Easy enough – let's translate the two values:
- _nein_  ⇔ no
- _ja_ ⇔ yes

In [35]:
# Creating a mapping dictionary to convert German unrepaired damage types to English
unrepaired_damage_mapping = {
    "nein": "no",
    "ja": "yes"
}

# Mapping the German unrepaired damage types to English
autos["unrepaired_damage"] = autos["unrepaired_damage"].map(unrepaired_damage_mapping)

# Verifying the changes
autos["unrepaired_damage"].value_counts()

unrepaired_damage
no     32071
yes     4197
Name: count, dtype: int64

## Exploring the date columns
As mentioned before, the `date_crawled`, `ad_created` and `ad_last_seen` columns are formatted into string objects. To make sense of them quantitavely and be able to use them in analyses, we want to convert them into numeric values.

Note that we're not converting the integers in `registration_month` and `registration_year` here. Reminder: `TimeStamp` (like `DateTime`) objects require a full date consisting of days, months and years. That makes these columns' values unsuitable for conversion.

Before all else, let's explore the dates and times listed in the `date_crawled`, `ad_created` and `ad_last_seen` columns and check out their current formatting.

In [36]:
# Printing the first five date strings in each column
autos[["date_crawled", "ad_created", "ad_last_seen"]].head()

date_crawled           ad_created         ad_last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
5  2016-03-21 13:47:45  2016-03-21 00:00:00  2016-04-06 09:45:21

Above, we can see that the strings contain both dates and times. The first ten characters contain encompass the dates.

Let's convert these columns to `DateTime` objects.

### Converting `date_crawled`, `ad_created` and `ad_last_seen` to `DateTime` objects

In [37]:
autos["date_crawled"] = pd.to_datetime(autos["date_crawled"], format="%Y-%m-%d %H:%M:%S")
autos["ad_created"] = pd.to_datetime(autos["ad_created"], format="%Y-%m-%d %H:%M:%S")
autos["ad_last_seen"] = pd.to_datetime(autos["ad_last_seen"], format="%Y-%m-%d %H:%M:%S")

# Verifying the changes
autos[["date_crawled", "ad_created", "ad_last_seen"]].info()

<class 'pandas.core.frame.DataFrame'>
Index: 42558 entries, 0 to 49999
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   date_crawled  42558 non-null  datetime64[ns]
 1   ad_created    42558 non-null  datetime64[ns]
 2   ad_last_seen  42558 non-null  datetime64[ns]
dtypes: datetime64[ns](3)
memory usage: 1.3 MB


### Exploring distributions of `date_crawled`, `ad_created` and `ad_last_seen`
Because we're mainly interested in the dates, we'll create a distribution (using `.value_counts()`) using only the first ten characters from the strings. That will also make the value counts more manageable, given the large number of times these value counts would otherwise include!

To do this, we use the 'floor' (lowest 'part' of a `TimeFrame` object) to demarcate what parts of the object we want to include: from the floor up. If we use the day as the floor, Python will know to include the days, months and years in its calculations.

In [38]:
# Exploring the date_crawled column – assessing the percentages of unique dates to determine the distribution
autos["date_crawled"].dt.floor(freq="d").value_counts(normalize=True, dropna=False).sort_index()

date_crawled
2016-03-05   0.0254
2016-03-06   0.0141
2016-03-07   0.0365
2016-03-08   0.0335
2016-03-09   0.0332
2016-03-10   0.0321
2016-03-11   0.0321
2016-03-12   0.0374
2016-03-13   0.0157
2016-03-14   0.0368
2016-03-15   0.0340
2016-03-16   0.0293
2016-03-17   0.0313
2016-03-18   0.0127
2016-03-19   0.0345
2016-03-20   0.0385
2016-03-21   0.0370
2016-03-22   0.0324
2016-03-23   0.0317
2016-03-24   0.0292
2016-03-25   0.0319
2016-03-26   0.0326
2016-03-27   0.0305
2016-03-28   0.0350
2016-03-29   0.0339
2016-03-30   0.0338
2016-03-31   0.0319
2016-04-01   0.0340
2016-04-02   0.0354
2016-04-03   0.0392
2016-04-04   0.0369
2016-04-05   0.0129
2016-04-06   0.0031
2016-04-07   0.0014
Name: proportion, dtype: float64

This shows us that the ads were crawled over a period of just a little over one month.

Now let's see what we can glean from the distribution of the `ad_created` column.

In [39]:
# Exploring the ad_created column – assessing the percentages of unique values to determine the distribution
autos["ad_created"].dt.floor(freq="d").value_counts(normalize=True, dropna=False).sort_index()

ad_created
2015-08-10   0.0000
2015-09-09   0.0000
2015-11-10   0.0000
2015-12-05   0.0000
2015-12-30   0.0000
              ...  
2016-04-03   0.0395
2016-04-04   0.0373
2016-04-05   0.0116
2016-04-06   0.0031
2016-04-07   0.0013
Name: proportion, Length: 72, dtype: float64

The ads crawled for this dataset, were created between early august of 2015, and early april of 2016. That would mean that on average, about 205 ads were created per day (50000 ads / 244 days). Impressive!

Granted, this does not take into account several issues in this dataset, such as the fact that there is a near month-long gap between the first two values in the value count above. Could there really have been a whole month without any ad creations?

To be sure, let's see the distribution of created ads between the two years.

In [40]:
# Exploring the ad_created column – assessing the percentages of unique years to determine the distribution
autos["ad_created"].dt.year.value_counts(normalize=True, dropna=False).sort_index()

ad_created
2015   0.0001
2016   0.9999
Name: proportion, dtype: float64

This shows us that the grand majority of ads that were crawled for this dataset, were created in 2016. Given that the last date in `ad_created` is the 7th of April of 2016, that would turn the average number of ads per day into something closer to 50000 ads / 97 days ≈ 515 ads a day.

This does beg the question as to why hardly any ads were created in the final four months of 2015. These numbers, however, may be a result of a fault in the dataset.

Let's also check out at what time people tend to create their listings.

In [41]:
# Exploring the ad_created column – assessing the numbers of unique hours to determine the distribution
autos["ad_created"].dt.time.value_counts(normalize=True, dropna=False).sort_index()

ad_created
00:00:00   1.0000
Name: proportion, dtype: float64

This makes it seem like all listings were created exactly at midnight, which doesn't make sense. We can assume this is the result of the crawler only having access to the date of creation, or a choice made by the person who collected the data not to include detailed time information. Unfortunately, this means we can't come to any meaningful conclusions about the times at which listings are most often created.

Let's now check out `ad_last_seen`'s distribution.

In [42]:
# Exploring the ad_last_seen column – assessing the percentages of unique values to determine the distribution
autos["ad_last_seen"].dt.floor(freq="d").value_counts(normalize=True, dropna=False).sort_index()

ad_last_seen
2016-03-05   0.0011
2016-03-06   0.0040
2016-03-07   0.0050
2016-03-08   0.0072
2016-03-09   0.0097
2016-03-10   0.0104
2016-03-11   0.0123
2016-03-12   0.0236
2016-03-13   0.0085
2016-03-14   0.0128
2016-03-15   0.0159
2016-03-16   0.0159
2016-03-17   0.0276
2016-03-18   0.0071
2016-03-19   0.0157
2016-03-20   0.0206
2016-03-21   0.0200
2016-03-22   0.0207
2016-03-23   0.0185
2016-03-24   0.0192
2016-03-25   0.0189
2016-03-26   0.0165
2016-03-27   0.0156
2016-03-28   0.0204
2016-03-29   0.0219
2016-03-30   0.0249
2016-03-31   0.0234
2016-04-01   0.0233
2016-04-02   0.0249
2016-04-03   0.0250
2016-04-04   0.0241
2016-04-05   0.1265
2016-04-06   0.2261
2016-04-07   0.1328
Name: proportion, dtype: float64

All the ads were last seen within a month of each other, apparently: starting at the 5th of March of 2016, and ending at the 7th of April of 2016. Moreover, over half of the ads were seen during the period of the last three days.

This information can be used to determine when an entry was removed (e.g. the car being sold, the listing expiring or the listing being taken down for any other reason). However, when cross-referencing the rows (especially of those whose ads were last seen during the last 3 days) with those of the `date_crawled` column, we might assume that those numbers are so high simply because those dates mark the end of the crawling time window.

## Analysing the average price per brand
With the data cleaning out of the way, it's time for some actual analyses. Let's take a look at prices per brands, and how the two might relate to each other. To do this, we'll perform aggregation by the `brand` column to explore differences in price across brands.

First, however, let's explore the `brand` column.

### Exploring the `brand` column

In [43]:
explore_column(autos, "brand")

Exploring the brand column – assessing the numbers of unique values, description, and value counts
Unique shape: 
 (40,)
Description: 
 count          42558
unique            40
top       volkswagen
freq            9136
Name: brand, dtype: object
Value counts, sorted ascending: 
 brand
alfa_romeo        288
audi             3755
bmw              4846
chevrolet         235
chrysler          157
citroen           590
dacia             114
daewoo             55
daihatsu           99
fiat             1072
ford             2923
honda             346
hyundai           434
jaguar             64
jeep               93
kia               310
lada               21
lancia             45
land_rover         74
mazda             660
mercedes_benz    4046
mini              399
mitsubishi        349
nissan            631
opel             4576
other cars        316
peugeot          1272
porsche           141
renault          1927
rover              51
saab               72
seat              796
skoda    

Here, we can see we have a total of 40 unique car brands in this column, across 44001 rows. However, the distribution is not at all equal. Let's look at percentages to get a better idea:

In [44]:
autos["brand"].value_counts(normalize=True)

brand
volkswagen      0.2147
bmw             0.1139
opel            0.1075
mercedes_benz   0.0951
audi            0.0882
ford            0.0687
renault         0.0453
peugeot         0.0299
fiat            0.0252
seat            0.0187
skoda           0.0170
mazda           0.0155
nissan          0.0148
citroen         0.0139
smart           0.0139
toyota          0.0129
hyundai         0.0102
volvo           0.0095
mini            0.0094
mitsubishi      0.0082
honda           0.0081
other cars      0.0074
kia             0.0073
alfa_romeo      0.0068
suzuki          0.0060
chevrolet       0.0055
chrysler        0.0037
porsche         0.0033
dacia           0.0027
daihatsu        0.0023
subaru          0.0023
jeep            0.0022
land_rover      0.0017
saab            0.0017
jaguar          0.0015
daewoo          0.0013
rover           0.0012
trabant         0.0011
lancia          0.0011
lada            0.0005
Name: proportion, dtype: float64

Now we can see that **Volkswagen cars** take up **almost 22%** of all cars in terms of brand. The #2 most popular, BMW, takes up a little over half of that, at about 11%. 

However, _'other_cars'_ is a category on eBay that includes various brands of cars. Therefore, we'll want to exclude rows with this value from further analyses:

In [45]:
filtered_autos_brands = autos[autos["brand"] != "other_cars"]

 Let's see what percentage of cars the ten most popular brands account for, without including `other_cars`.

In [46]:
top_10_brand_proportion = filtered_autos_brands["brand"].value_counts(normalize=True).head(10).sum()
print(top_10_brand_proportion)

0.8071102965364912


So, the top 10 brands amounts to **81%** of all brands, so we will focus on these brands moving forward. Over half of these brands are German or use German engines, at least, and all 10 brands are European in origin. This makes sense, given that Germany is a European country.

Aggregation will occur across all brands for completeness' sake, but the majority of consequent analysis will be focussed on the top 10 brands.

In [47]:
# Saving the top 10 brands to a variable, in case we wind up needing it later
top_10_brands = filtered_autos_brands["brand"].value_counts().head(10).index

### Aggregating average price per brand
Let's now find out what the average price per brand is.

In [48]:
# Creating an empty dictionary to hold aggregate data
price_per_brand = {}

# Looping through the top 10 brands and calculating the mean price for each
for brand in top_10_brands:
    brand_only = autos[autos["brand"] == brand] # Filtering the DataFrame for the current brand
    mean_price = brand_only["price"].mean() # Calculating the mean price for the current brand
    price_per_brand[brand] = int(mean_price) # Adding the mean price to the dictionary

# Converting the dictionary to a Series and sorting it in descending order
price_per_brand_series = pd.Series(price_per_brand).sort_values(ascending=False)

print(price_per_brand_series)

audi             8599
mercedes_benz    8002
bmw              7869
volkswagen       5435
seat             4586
ford             3756
peugeot          3201
opel             3062
fiat             2892
renault          2592
dtype: int64


Now we can see that **audi** is the priciest brand among the top 10 most popular brands during the time window of early August 2015 to early April 2016, peaking at a mean price of $8503 per car, with **Renault** being the cheapest on average at $2558. Meanwhile, **Volkswagen** sits around the middle, averaging at $5369,92.

Let's now get an idea of the average mileage of cars belonging to the top 10 most popular brands, and then combine the consequent Series object with the `price_per_brand_series` to see how mileage affects the price (and, who knows, maybe vice versa).

## Analysing average mileage per brand

In [49]:
# Creating an empty dictionary to hold aggregate data
mileage_per_brand = {}

# Looping through the top 10 brands and calculating the mean mileage for each
for brand in top_10_brands:
    brand_only = autos[autos["brand"] == brand] # Filtering the DataFrame for the current brand
    mean_mileage = brand_only["odometer_km"].mean() # Calculating the mean mileage for the current brand
    mileage_per_brand[brand] = int(mean_mileage) # Adding the mean mileage to the dictionary

# Converting the dictionary to a Series and sorting it in descending order
mileage_per_brand_series = pd.Series(mileage_per_brand).sort_values(ascending=False)

print(mileage_per_brand_series)

bmw              133690
mercedes_benz    132090
audi             130881
opel             129229
volkswagen       128648
renault          127101
peugeot          126312
ford             124526
seat             120835
fiat             116511
dtype: int64


We can now see that there is a 2000 kilometer gap between a _Land Rover's_ average mileage (134391,89 km), and that of a _Mini_ (89399,51 km).

What does this mean in relation to the price, however? What happens to price as mileage goes up among the top 10 most popular brands, and vice versa?

Let's turn the two series into a single DataFrame, and see for ourselves.

### Creating new DataFrame: most popular brands and their avg mileages and avg prices

In [50]:
# Create a new DataFrame from the two series
brand_info = pd.DataFrame({
    "Average Mileage": mileage_per_brand_series,
    "Average Price": price_per_brand_series
})

# Sort the DataFrame by "Average Price" in descending order
brand_info_sorted = brand_info.sort_values(by="Average Mileage", ascending=True)

# Display the sorted DataFrame
print(brand_info_sorted)

               Average Mileage  Average Price
fiat                    116511           2892
seat                    120835           4586
ford                    124526           3756
peugeot                 126312           3201
renault                 127101           2592
volkswagen              128648           5435
opel                    129229           3062
audi                    130881           8599
mercedes_benz           132090           8002
bmw                     133690           7869


The range of mileages doesn't vary so much in the top 10 most popular brands, with 116805 kilometers for _Fiat_, and 133799 km for _BMW_. That puts them only about 2000 kilometers apart.

Interestingly enough, Fiat is the second-to-lowest priced brand on average in this DataFrame (at about $2895 on average), while BMW is the third most expensive on average at $7816. 

## Analysing popular brand and vehicle type combinations
Now, it might be interesting to get an idea of what vehicle types are popular as well, and especially in combination with what brands. This would be particularly useful for people who sell cars for a living – to assess 'competition' on eBay, and potentially get an idea for the existing supply and find gaps therein.

For brevity's sake, we'll use the `top_10_brands` variable, rather than looking at all 40 brands in the dataset (or 39, if we're leaving `other_cars` aside).

In [51]:
# Creating an empty dictionary to hold aggregate data
vehicle_type_per_brand = {}

# Looping through the top 10 brands and calculating the most common vehicle type for each
for brand in top_10_brands:
    brand_only = autos[autos["brand"] == brand]  # Filtering the DataFrame for the current brand
    vehicle_type_mode = brand_only["vehicle_type"].mode()  # Calculating the most common vehicle type for the current brand
    vehicle_type_per_brand[brand] = vehicle_type_mode[0]  # Adding the most common vehicle type to the dictionary

# Converting the dictionary to a Series and sorting it
vehicle_type_per_brand_series = pd.Series(vehicle_type_per_brand)

print(vehicle_type_per_brand_series)

volkswagen               sedan
bmw                      sedan
opel                 supermini
mercedes_benz            sedan
audi             station wagon
ford                 supermini
renault              supermini
peugeot              supermini
fiat                 supermini
seat                 supermini
dtype: object


The clear 'winner' here is the supermini: a compact type of car often seen in urban areas in Europe, because it's easy to move around the tight streets of old cities and villages.

## Analysing the relation between mileage and price
To make the data in the `odometer_km` a little more easy to manage, we'll first split it into groups. After that, we'll use the `groupby()` method to group the price data by these odometer groups.

Let's first check out what the distribution of the `odometer_km` values looks like.

In [52]:
autos["odometer_km"].describe()

count    42558.0000
mean    126242.0696
std      38799.6318
min       5000.0000
25%     125000.0000
50%     150000.0000
75%     150000.0000
max     150000.0000
Name: odometer_km, dtype: float64

Based on this, we might want to divide odometer data as follows:

In [53]:
# Defining the odometer_km ranges to group the price data by
bins = [0, 20000, 40000, 60000, 80000, 100000, 120000, 140000, 160000]
labels = ["0-20k", "20k-40k", "40k-60k", "60k-80k", "80k-100k", "100k-120k", "120k-140k", "140k-160k"]

# Categorising the odometer_km data into the defined ranges
autos["odometer_group"] = pd.cut(autos["odometer_km"], bins=bins, labels=labels)

# Grouping the prices by odometer_group and calculating the average price for each group
average_prices_per_group = autos.groupby("odometer_group")["price"].mean()

# Displaying the average prices per odometer_group
print(average_prices_per_group)

odometer_group
0-20k       11351.9048
20k-40k     13363.0753
40k-60k     11483.4460
60k-80k      9795.5173
80k-100k     7897.6488
100k-120k          NaN
120k-140k    5967.4538
140k-160k    3795.2401
Name: price, dtype: float64


/var/folders/z4/3sk68njj2_s65s5s73mlpdsr0000gn/T/ipykernel_14660/1431629638.py:9: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  average_prices_per_group = autos.groupby("odometer_group")["price"].mean()


Interestingly, it seems there are no data points in the 100k-120k range. Let's see if this has to do with missing data points (i.e. no cars labelled between these prices):

In [54]:
autos["odometer_group"].value_counts()

odometer_group
140k-160k    27465
120k-140k     4559
80k-100k      3425
60k-80k       2399
40k-60k       1971
20k-40k       1394
0-20k         1345
100k-120k        0
Name: count, dtype: int64

Though surprising, it isn't entirely impossible that there should be no car with 100k - 120k km on their odometer. These cars might have sold faster than the crawler could keep up with, as these price ranges often present a very appealing price-to-quality ratio: cars within this range of kilometers on the odometer tend to be cheap, but not driven to the point of wear-and-tear becoming a big issue.

Either way, looking at the original print-out of `average_prices_per_group`, we can see the following:

- The price follows the expected trend: as the number of kilometers on the odometer go up, the price goes down. After 60k, and even more so after 80k, prices start to plummet.
- Interestingly, between 0 - 60k kilometers, the prices don't fluctuate much.
- In fact, between 0 - 20k kilometers on the odometer costs less than cars with between 20k - 40k.

## Analysing the relation between price and unrepaired damage
How much cheaper are cars with damage than their non-damaged counterparts (on average)? We can find out by comparing the mean prices of cars with unrepaired damage, and the mean prices of those without unrepaired damage.

In [55]:
# Grouping the data by unrepaired_damage and saving the groups to separate variables
damaged = autos[autos["unrepaired_damage"] == "yes"]
undamaged = autos[autos["unrepaired_damage"] == "no"]

# Calculating the average price for damaged and undamaged cars
average_price_damaged = damaged["price"].mean()
average_price_undamaged = undamaged["price"].mean()

# Displaying the average prices for damaged and undamaged cars, with 2 decimal places
print(f"Average price for damaged cars: ${average_price_damaged:.2f}")
print(f"Average price for undamaged cars: ${average_price_undamaged:.2f}")

Average price for damaged cars: $2186.40
Average price for undamaged cars: $6524.28


That is a pretty stark difference. Undamaged cars are about three times more expensive on average than damaged ones. However, the CSV file contains no information about the nature of the damage. As such, this column likely includes cars that are being sold for parts or have other extensive damage that car enthusiasts might want to fix themselves – as well as cars with comparatively minor damage, such as scratches.

As such, it's hard to determine whether a seller will get the most out of their car sale with or without first repairing any damages.

## Conclusion

In this analysis, we worked with a dataset containing information about used cars. We performed data cleaning and analysis to gain insights into the dataset.

First, we cleaned the dataset by removing irrelevant columns and handling missing values. We also converted some columns to the appropriate data types.

Next, we analyzed the dataset to understand the distribution of car brands. We found that the top 10 most popular brands accounted for 81% of all brands. Among these brands, Audi had the highest average price, while Renault had the lowest average price.

We also explored various relations between various data points.

1.  between mileage and price. 
    - There was a slight variation in mileage among the top 10 popular brands, with Fiat having the lowest average mileage and BMW having the highest.
    - This type of information is useful for those looking for cars with the lowest possible mileage.

2. Popular brand and vehicle type combinations:
    - The most common vehicle type among the top 10 most popular brands is the supermini, followed by sedans and vans.
    - This information can be useful for car sellers to understand the competition on platforms like eBay and assess the existing supply in the market.

3. Relation between mileage and price:
    - There is a clear relationship between mileage and price. As the number of kilometers on the odometer increases, the price tends to decrease.
    - Between 0-60k kilometers, the prices don't fluctuate much, but after 60k and especially after 80k, prices start to decrease significantly.
    - This information can be valuable for both car buyers and sellers. Buyers can use it to negotiate prices based on mileage, and sellers can set competitive prices based on the mileage of their vehicles.

4. Relation between price and unrepaired damage:
    - Cars with unrepaired damage are significantly cheaper on average compared to undamaged cars.
    - The average price of damaged cars is approximately 3 times lower than that of undamaged cars.
    - However, it's important to note that the dataset does not provide information about the nature or extent of the damage.
    - Sellers should consider whether repairing the damage before selling will result in a higher selling price, as some buyers may prefer to purchase undamaged cars.

These analyses provide valuable insights into the used car market in Germany. The information can be used by car buyers and sellers to make informed decisions about pricing, brand preferences, and the impact of mileage and damage on prices. Additionally, researchers and analysts can utilize this data to study market trends and patterns in the automotive industry.